PSA return status coverage script

gilles-peskine-arm opened the pull request:
Add infrastructure to run unit tests and collect the return values for every PSA API function that returns psa_status_t.

    ./tests/scripts/psa_collect_statuses.py >statuses.txt

Status: runs ok but not very well documented and the build script is not very robust.

Patater commented:
This may conflict with https://github.com/ARMmbed/mbed-crypto/pull/213

I like the idea of checking that return codes are documented. I think the API is stable enough now that having that capability will be a help and not a hindrance.

Is it possible to continue work on this?

gilles-peskine-arm commented:
I'll rebase this on top of #213 if we decide to adopt it.

Are you ok with the current design? It was good enough for my goal at the time, which was to compare reachable errors with documented errors and is still relevant. But it isn't very well integrated into Mbed Crypto. The script I have for the other side, to collect documented errors, works from the PSA spec (private link: https://github.com/ARMmbed/psa-crypto/pull/227).

Patater commented:
> Are you ok with the current design?

It's a bit funky in that its a mix of shell in a Makefile and some Python, but it's beneficial as is. We can refactor into something better as needed.

gilles-peskine-arm commented:
I rebased the changes on top of current `development`. I added some documentation, but I did not change the design. I think this is good enough to include, and we'll see later whether to improve the design or to leave it as a minor hack depending on how useful we think this is.

Patater commented on tests/scripts/all.sh:
```diff
@@ -629,6 +629,16 @@ component_test_everest () {
+component_test_psa_collect_statuses () {
+  msg "build+test: psa_collect_statuses" # ~30s
+  scripts/config.pl full
+  scripts/config.pl unset MBEDTLS_MEMORY_BUFFER_ALLOC_C # slow and irrelevant
```
We may not need to add `unset MBEDTLS_MEMORY_BUFFER_ALLOC_C` if we sideport the removal of MBEDTLS_MEMORY_BUFFER_ALLOC_C from full config. Necessary for now.

gilles-peskine-arm replied:
We definitely should sideport that change. I'd expect it to come around the next time we sync from mbedtls.

k-stachowiak commented on tests/scripts/psa_collect_statuses.py:
```diff
@@ -0,0 +1,125 @@
+            subprocess.check_call(['make', 'clean'],
+                                  cwd='tests',
+                                  stdout=sys.stderr)
+        with open(os.devnull, 'w') as devnull:
```
Why not store the build log in case of a failure?

gilles-peskine-arm replied:
Good idea. I wrote this quickly and primarily for interactive use where I didn't want to scroll through build logs, but the log should definitely be stored in a file. https://github.com/ARMmbed/mbed-crypto/issues/252
